### PR TITLE
Incident edges accessor

### PR DESCRIPTION
**Pull request by d-michail**
A proposal for issue #298, trying to implement the discussion there. 

@jsichi, @jkinable Comments are welcome, regarding both functionality and naming and/or location of methods, factories, etc. 

I only applied the proposed pattern to the `BidirectionalDijkstraShortestPath` class for demonstration purposes. There are multiple places in the library that it could/should be applied.

### CODE REVIEW

**Comment by jsichi** [comment]
Looks fine to me!

**Comment by jkinable** [comment]
I'm currently traveling within the US; will be back early next week. Would like to have a look at this.

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java [hunk 1]
```diff
@@ -547,6 +547,65 @@
...
+     * outgoing/incoming edges of a vertex. In undirected graphs or if the graph is neither directed
+     * or undirected, the accessor returns all incident edges of a vertex.
+     *
+     * @param graph the input graph
+     * @return an vertex edge accessor for the incoming/outgoing edges
+     *
+     * @param <V> the graph vertex type
+     * @param <E> the graph edge type
+     */
+    public static <V, E> IncidentEdgeAccessor<V, E> incomingOutgoingEdgeAccessor(Graph<V, E> graph)
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: Just a thought (feel free to reject): what about naming the methods:
- incomingEdgesOfAccessor
- outgoingEdgesOfAccessor
- edgesOfAccessor

An alternative name for `incomingOutgoingEdgeAccessor` could be simply: `incidentEdgesAccessor`

File: jgrapht-core/src/main/java/org/jgrapht/IncidentEdgeAccessor.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht;
+
+/**
+ * A vertex incident edges accessor. Incident edge accessors are helpful in creating algorithms
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/IncidentEdgeAccessor.java]: Maybe we need to be a bit more verbose in the description. At least, based on the current description, I wouldn't really know what to do with this interface. Here's one attempt.

An accessor for edges incident to a vertex. Incident edge accessors are helpful in creating algorithms
which operate on both directed and undirected graphs. Through the accessor, one can iterate over a (subset of) the edges incident to a particular vertex; the exact implementation of the accessor is often dependent on the underlying graph (directed or undirected).

**Comment by d-michail** [comment]
I also prefer your suggestion. 

I would also move upwards the degree functions from both `DirectedGraph` and `UndirectedGraph`. These interfaces should only function as markers in order to understand the behavior of the methods. The change will not be very invasive as the implementation class `AbstractBaseGraph` already does that.

**Comment by jsichi** [comment]
Wow, we started with a minor refactoring and now we're discussing revamping the core interfaces :)

I think things are actually fine as they are, and it's good to keep a separation between fundamental graph structure and how it's accessed (which is why I like the pull request as is).

However, if you both really want to go that far, then I would say be even more radical and get rid of the Directed/Undirected interfaces entirely, along with everything that implies (e.g. allow graphs which mix directed with undirected edges, which has been a recurring request).

**Pull request closed without merging**